#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QDebug>
#include <QStack>
#include <QMessageBox>

MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent), ui(new Ui::MainWindow)
{
    ui->setupUi(this);
}

MainWindow::~MainWindow()
{
    delete ui;
}

QHash<QChar, double> map;
QStack<double> values;
QStack<QChar> operators;

bool isOperator(QChar c)
{
    return (c == '+' || c == '-' || c == '*' || c == '/');
}

int operatorPriority(QChar c)
{
    if(c == '+' || c == '-')
        return 0;

    if(c == '*' || c == '/')
        return 1;

    return -1;
}

bool MainWindow::proceedOperation()
{
    double b = values.top(); // Возвращает ссылку на верхний элемент стека
    values.pop(); // Удаляет верхний элемент из стека и возвращает его
    double a = values.top();
    values.pop();

    QChar oper = operators.top();
    operators.pop();

    double result;

    switch(oper.unicode())
    {
    case '+':
        result = a + b;
        break;
    case '-':
        result = a - b;
        break;
    case '*':
        result = a * b;
        break;
    case '/':
    {
        if(b == 0)
        {
            QMessageBox::warning(this, "Invalid operation", "Division by zero is not possible");
            return false;
        }
        result = a / b;
        break;
    }
    default:
        result = 0;
    }

    values.push(result); // Добавляет элемент result в начало стека.

    return true;
}

void MainWindow::on_pushButtonCalculate_clicked()
{
    map.insert('a', ui->lineEditA->text().toDouble());
    map.insert('b', ui->lineEditB->text().toDouble());
    map.insert('c', ui->lineEditC->text().toDouble());
    map.insert('d', ui->lineEditD->text().toDouble());
    map.insert('e', ui->lineEditE->text().toDouble());

    QString expression = ui->lineEditExpression->text();

    for (int i = 0; i < expression.length(); i++)
    {
        if (expression[i] == ' ')
            continue;

        if (isOperator(expression[i]))
        {
            while (!operators.empty() && operatorPriority(expression[i]) < operatorPriority(operators.top()))
            {
                if (!proceedOperation())
                    return;
            }
            operators.push(expression[i]);
        }
        else if (expression[i] == '(')
            operators.push(expression[i]);
        else if (expression[i] == ')')
        {
            while (operators.top() != '(')
            {
                if (!proceedOperation())
                    return;
            }
            operators.pop();
        }
        else
            values.push(map.value(expression[i]));
    }

    while (!operators.empty())
    {
        if (!proceedOperation())
            return;
    }

    ui->lineEditResult->setText(QString::number(values.top()));
    values.pop();
}

void MainWindow::on_lineEditExpression_textChanged()
{
    if (ui->lineEditExpression->text() != "")
        ui->pushButtonCalculate->setEnabled(true);
    else
        ui->pushButtonCalculate->setEnabled(false);
}

void MainWindow::on_lineEditA_textChanged()
{
    if (ui->lineEditA->text() != "")
        ui->pushButtonCalculate->setEnabled(true);
    else
        ui->pushButtonCalculate->setEnabled(false);
}

void MainWindow::on_lineEditB_textChanged()
{
    if (ui->lineEditB->text() != "")
        ui->pushButtonCalculate->setEnabled(true);
    else
        ui->pushButtonCalculate->setEnabled(false);
}

void MainWindow::on_lineEditC_textChanged()
{
    if (ui->lineEditC->text() != "")
        ui->pushButtonCalculate->setEnabled(true);
    else
        ui->pushButtonCalculate->setEnabled(false);
}

void MainWindow::on_lineEditD_textChanged()
{
    if (ui->lineEditD->text() != "")
        ui->pushButtonCalculate->setEnabled(true);
    else
        ui->pushButtonCalculate->setEnabled(false);
}

void MainWindow::on_lineEditE_textChanged()
{
    if (ui->lineEditE->text() != "")
        ui->pushButtonCalculate->setEnabled(true);
    else
        ui->pushButtonCalculate->setEnabled(false);
}
